### PR TITLE
sendsig() for native purecap

### PR DESCRIPTION
**Pull request by bsdjhb**
This finishes enabling the remaining bits of the sysvec for native purecap.  There is still much to do with updating native syscalls to work with purecap which will also spill over into mcontext_t, etc. but this should finish off elf_machdep.c.

### CODE REVIEW

File: sys/mips/mips/freebsd64_machdep.c [hunk 1]
```diff
@@ -198,7 +200,224 @@ static void
...
+			sigexit(td, SIGILL);
+			/* NOTREACHED */
+		}
+#endif
+		sp = (vm_offset_t)regs->sp;
+	}
+#ifdef CPU_CHERI
+	cp2_len = sizeof(*cfp);
+	sp -= cp2_len;
+	sp &= ~(CHERICAP_SIZE - 1);
```
Comment (arichardson) [inline sys/mips/mips/freebsd64_machdep.c]: Probably copied code but I would use:
```suggestion
	sp = rounddown2(sp, CHERICAP_SIZE);
```

File: sys/mips/mips/pm_machdep.c [hunk 2]
```diff
@@ -180,15 +187,21 @@ sendsig(sig_t catcher, ksiginfo_t *ksi, sigset_t *mask)
 #endif
 
 	/* save user context */
-	bzero(&sf, sizeof(sf));
+	bzero(&sf, sizeof(struct sigframe));
```
Comment (arichardson) [inline sys/mips/mips/pm_machdep.c]: Why this change? Using the variable seems less likely to break in the future.

File: sys/mips/mips/freebsd64_machdep.c [hunk 3]
```diff
@@ -198,7 +200,224 @@ static void
...
+	/*
+	 * Install CHERI signal-delivery register state for handler to run
+	 * in.  As we don't install this in the CHERI frame on the user stack,
+	 * it will be (genrally) be removed automatically on sigreturn().
+	 */
+	hybridabi_sendsig(td);
+#endif
+
+	regs->pc = (trapf_pc_t)catcher;
+	// FIXME: should this be an offset relative to PCC or an address?
```
Comment (arichardson) [inline sys/mips/mips/freebsd64_machdep.c]: Now that regs->pc is $pcc, I guess this FIXME can go. 
$t9 should match $pc which is the offset of $pcc.

File: sys/mips/mips/pm_machdep.c [hunk 4]
```diff
@@ -295,18 +313,30 @@ sendsig(sig_t catcher, ksiginfo_t *ksi, sigset_t *mask)
...
+	 */
+	if (cheri_is_sandboxed) {
+		regs->ddc = td->td_pcb->pcb_cherisignal.csig_ddc;
+		regs->idc = td->td_pcb->pcb_cherisignal.csig_idc;
+	}
+#else
 	regs->pc = (trapf_pc_t)catcher;
-	// FIXME: should this be an offset relative to PCC or an address?
-	regs->t9 = (register_t)(__cheri_offset intptr_t)regs->pc;
+	regs->t9 = (register_t)(intptr_t)catcher;
```
Comment (arichardson) [inline sys/mips/mips/pm_machdep.c]: For hybrid this should the offset of pcc/catcher.

File: sys/mips/mips/freebsd64_machdep.c (hunk 1 above)
Comment (bsdjhb) [reply sys/mips/mips/freebsd64_machdep.c]: Will do.

File: sys/mips/mips/pm_machdep.c (hunk 2 above)
Comment (bsdjhb) [reply sys/mips/mips/pm_machdep.c]: Reduces diff to upstream which uses the type.

File: sys/mips/mips/freebsd64_machdep.c (hunk 3 above)
Comment (bsdjhb) [reply sys/mips/mips/freebsd64_machdep.c]: Axed.

File: sys/mips/mips/pm_machdep.c (hunk 4 above)
Comment (bsdjhb) [reply sys/mips/mips/pm_machdep.c]: This is the no-CHERI-at-all-case, not hybrid.  hybrid is now in freebsd64_sendsig().

File: sys/mips/mips/pm_machdep.c [hunk 5]
```diff
@@ -228,35 +249,56 @@ sendsig(sig_t catcher, ksiginfo_t *ksi, sigset_t *mask)
...
 	} else {
 		/* Old FreeBSD-style arguments. */
 		regs->a1 = ksi->ksi_code;
+#if __has_feature(capabilities)
+		/*
+		 * XXX: Should this strip tags?
+		 */
+		regs->c4 = ksi->ksi_addr;
+#else
 		regs->a3 = (__cheri_addr uintptr_t)ksi->ksi_addr;
```
Comment (brooksdavis) [inline sys/mips/mips/pm_machdep.c]: Why the __cheri_addr in the non-CHERI case?

Comment (bsdjhb) [reply sys/mips/mips/pm_machdep.c]: I had left it from the hybrid code, but I've changed it to reduce the diff with upstream.

File: sys/mips/mips/freebsd64_machdep.c [hunk 6]
```diff
@@ -198,7 +200,223 @@ static void
...
+		sp = (vm_offset_t)regs->sp;
+	}
+#ifdef CPU_CHERI
+	cp2_len = sizeof(*cfp);
+	sp -= cp2_len;
+	sp = rounddown2(sp, CHERICAP_SIZE);
+	sf.sf_uc.uc_mcontext.mc_cp2state = sp;
+	sf.sf_uc.uc_mcontext.mc_cp2state_len = cp2_len;
+#endif
+	sp -= sizeof(struct sigframe);
```
Comment (jrtc27) [inline sys/mips/mips/freebsd64_machdep.c]: sigframe64

Comment (arichardson) [reply sys/mips/mips/freebsd64_machdep.c]: Or `sf`?

Comment (bsdjhb) [reply sys/mips/mips/freebsd64_machdep.c]: Fixed to use `sf`

**Pull request merged**